source/dofs/dof_handler.cc

```diff
@@ -2689,66 +2686,125 @@ DoFHandler<dim, spacedim>::prepare_coarsening_and_refinement(
   for (unsigned int i = 0; i < fe_index_for_hierarchy_level.size(); ++i)
     hierarchy_level_for_fe_index[fe_index_for_hierarchy_level[i]] = i;
 
+  //
+  // parallelization
+  //
+  // - create distributed copy of future fe index vector
+  // - update ghost values after each iteration (see later)
+  // - no need to compress, since the owning processor will have the correct FE
+  //   index
+
+  LinearAlgebra::distributed::Vector<int> future_fe_indices;
+  if (const auto parallel_tria =
+        dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+          &(*tria)))
+    {
+      const auto &partitioner =
+        parallel_tria->global_active_cell_index_partitioner();
+      future_fe_indices.reinit(partitioner.locally_owned_range(),
+                               partitioner.ghost_indices(),
+                               partitioner.get_mpi_communicator());
+    }
+  else
+    {
+      future_fe_indices.reinit(tria->n_active_cells());
+    }
+
+  // communicate active FE indices on ghost cells in case the user changed them
+  dealii::internal::hp::DoFHandlerImplementation::Implementation::
+    communicate_active_fe_indices(*this);
+
+  for (const auto &cell : active_cell_iterators())
+    if (cell->is_locally_owned())
+      future_fe_indices[cell->global_active_cell_index()] =
+        cell->future_fe_index();
+
   //
   // limit level difference of neighboring cells
   //
+  // - go over all locally relevant cells, and adjust the future FE indices of
+  //   locally owned neighbors to match the level difference (as a consequence,
+  //   indices on ghost cells will be updated only on the owning processor)
   // - always raise levels to match criterion, never lower them
-  //   (hence iterating from highest to lowest level)
-  // - update future FE indices
+  // - exchange future FE indices on ghost cells
 
   bool fe_indices_changed = false;
   bool fe_indices_changed_in_cycle;
   do
     {
       fe_indices_changed_in_cycle = false;
 
+      future_fe_indices.update_ghost_values();
+
       for (const auto &cell : active_cell_iterators())
-        {
-          const auto cell_fe_and_level =
-            hierarchy_level_for_fe_index.find(cell->future_fe_index());
+        if (!cell->is_artificial())
+          {
+            const auto cell_fe_and_level = hierarchy_level_for_fe_index.find(
+              future_fe_indices[cell->global_active_cell_index()]);
```

Comment:
I wouldn't search here. Bur rather fill the `future_fe_indices` vector already with normalized future-fe indices. This would mean that you have to translate at the beginning (before filling the vector) the future fe indices to the normalized indices and at the end (while setting the actual future-fe index the via `cell->set_future_fe_index()`) you would need to translate the normalized value to the actual value.

Reply:
You mean to have a `future_level_indices` distributed vector instead, and then translate them back at the end of the loop while actually setting `cell->set_future_fe_index()`? I like this.